python/ray/ml/train/integrations/sklearn/sklearn_trainer.py

```diff
@@ -0,0 +1,133 @@
+from typing import Union, Callable, Dict, Iterable, Optional
+
+import numpy as np
+import pandas as pd
+from ray.ml.trainer import GenDataset
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.preprocessor import Preprocessor
+from ray.util import PublicAPI
+from ray.ml.trainer import Trainer
+from ray.ml.checkpoint import Checkpoint
+
+from sklearn.base import BaseEstimator
+from sklearn.model_selection import BaseCrossValidator
+
+# some nice properties of Sklearn models:
+# - they are always pickleable
+# - no logic is present in __init__
+
+ArrayType = Union[pd.DataFrame, np.ndarray]
+MetricType = Union[str, Callable[[BaseEstimator, ArrayType, ArrayType], float]]
+ScoringType = Union[MetricType, Iterable[MetricType], Dict[str, MetricType]]
+CVType = Union[int, Iterable, BaseCrossValidator]
+
+
+@PublicAPI(stability="alpha")
+class SklearnTrainer(Trainer):
+    """A Trainer for scikit-learn estimator training.
+
+    This Trainer runs the ``fit`` method of the given estimator in a
+    non-distributed manner on a single Ray Actor.
+
+    The ``n_jobs`` (or ``thread_count``) parameters will be set to match
```

Comment:
For my own understanding, would multithreading even be effective here if we have python GIL?

Reply:
Most of the time, this is either joblib or C-level parallelism (independent of GIL).

Reply:
From what I understand the default joblib parallelism uses multiprocessing, is that correct? In which case I don't think multiprocessing will work well inside the trainable/Ray actor.

Reply:
yes, that is correct. we will have to force a thread joblib backend for now. We will want to use the Ray one eventually, but I don't know yet how to solve the issue with resources (see conversation above).

Reply:
Got it yeah- but what I'm asking is for thread joblib, from what I'm reading online, it seems like it's useful if you're able to disable the GIL. Is that something that we have to do, or that the user will need to handle in their code?

Reply:
I don't think we can disable the GIL, right?

The bottom line here is that we will be able to make estimators that make use of non-joblib parallelism use the correct number of CPUs, but in the initial implementation, I don't anticipate any real parallelism for joblib-based estimators.

Reply:
I'll think about how to make ray joblib work here, though.

Reply:
+1 on joblib as future work. Maybe leave this as a TODO that true multi-processing is possible but not implemented yet? That clarifies to the user why this isn't offered.